src/java/com/unifina/signalpath/blockchain/GetEvents.java

```diff
@@ -103,17 +103,19 @@ static void convertAndSend(Output output, Object value) {
 	@Override
 	public void onEvent(JSONArray events) {
 		try {
-			String txHash = events.getJSONObject(0).getString("transactionHash");
-			txHashOutput.send(txHash);
-			log.info(String.format("Received event '%s'", txHash));
-			TransactionReceipt txr = Web3jHelper.waitForTransactionReceipt(web3j, txHash, check_result_waitms, check_result_max_tries);
-			if(txr == null){
-				log.error("Couldnt find TransactionReceipt for transaction "+txHash+ " in allotted time");
-				return;
+			int eventcount = events.length();
+			for(int i=0; i<eventcount; i++) {
+				String txHash = events.getJSONObject(i).getString("transactionHash");
+				txHashOutput.send(txHash);
+				log.info(String.format("Received event '%s'", txHash));
+				TransactionReceipt txr = Web3jHelper.waitForTransactionReceipt(web3j, txHash, check_result_waitms, check_result_max_tries);
+				if (txr == null) {
+					log.error("Couldnt find TransactionReceipt for transaction " + txHash + " in allotted time");
+					return;
+				}
+				displayEventsFromLogs(txr.getLogs());
+				asyncPropagator.propagate();
```

Comment:
@hpihkala @jtakalai  I don't exactly understand this asyncPropagator.propagate() call. I think Juuso added it in the stub he gave me. I'm still having an issue propagating all events from a single JSON response. When I run this normally, only the 1st value is propagated, but when I put a breakpoint here (line 117) and wait a few seconds, it propagates.

Reply:
@isentropy I think it's because the events aren't being added to the event queue and handled (and propagated) in the event queue consumer thread. How this is done is a bit messy currently, but it will actually become much cleaner in a refactor which is almost ready. Anyway, the current pattern is this:

- In the method called by your module-specific thread (here, `onEvent`), produce events to the DataSource event queue. Best pattern would probably be insert one `FeedEvent` per each event in the transaction.

```
DataSource dataSource = getGlobals().getDataSource();
for (/* each event in transaction */) {
    dataSource.enqueueEvent(new FeedEvent<>(/* pass as content whatever you need to handle the event, you might need to wrap it in ITimestamped interface */));
}
```

- Then implement a method called `receive(FeedEvent event)`, which gets called by the event loop once the event you inserted in the previous step comes up from the queue. Here, you should send values to outputs and then propagate.

```
        @Override
	public void receive(FeedEvent event) {
		if (event.content instanceof /* whatever you inserted */) {
			sendOutput(event.content);
			getPropagator().propagate();
		} else {
			super.receive(event);
		}
	}
```

Reply:
OK, I'm familiar with that pattern from SendEthereumTransaction. I will update